server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java

```diff
@@ -30,139 +28,129 @@
 
-            FieldFetcher nestedSubFieldFetcher = FieldFetcher.create(
-                context,
-                fieldAndFormats,
-                narrowedScopeNestedMappings,
-                nestedFieldPath
-            );
+        return new FieldFetcher(fieldContexts, unmappedFieldFetcher);
+    }
 
-            // add a special ValueFetcher that filters source and collects its subfields
-            fieldContexts.put(
-                nestedFieldPath,
-                new FieldContext(nestedFieldPath, new NestedValueFetcher(nestedFieldPath, nestedSubFieldFetcher))
-            );
+    private static UnmappedFieldFetcher buildUnmappedFieldFetcher(
+        SearchExecutionContext context,
+        Set<String> mappedFields,
+        String nestedScope,
+        List<String> unmappedFetchPatterns
+    ) {
+        if (unmappedFetchPatterns.isEmpty()) {
+            return UnmappedFieldFetcher.EMPTY;
         }
+        Set<String> mappedAndNestedFields = new HashSet<>(mappedFields);
+        mappedAndNestedFields.addAll(context.nestedLookup().getImmediateChildMappers(nestedScope));
+        return new UnmappedFieldFetcher(mappedAndNestedFields, unmappedFetchPatterns);
+    }
 
-        CharacterRunAutomaton unmappedFieldsFetchAutomaton = null;
-        // We separate the "include_unmapped" field patters with wildcards from the rest in order to use less
-        // space in the lookup automaton
-        Map<Boolean, List<String>> partitions = unmappedFetchPattern.stream()
-            .collect(Collectors.partitioningBy((s -> Regex.isSimpleMatchPattern(s))));
-        List<String> unmappedWildcardPattern = partitions.get(true);
-        List<String> unmappedConcreteFields = partitions.get(false);
-        if (unmappedWildcardPattern.isEmpty() == false) {
-            unmappedFieldsFetchAutomaton = new CharacterRunAutomaton(
-                Regex.simpleMatchToAutomaton(unmappedWildcardPattern.toArray(new String[unmappedWildcardPattern.size()])),
-                AUTOMATON_MAX_DETERMINIZED_STATES
-            );
+    private static ValueFetcher buildValueFetcher(SearchExecutionContext context, ResolvedField fieldAndFormat) {
+        try {
+            return fieldAndFormat.ft.valueFetcher(context, fieldAndFormat.format);
+        } catch (IllegalArgumentException e) {
+            StringBuilder error = new StringBuilder("error fetching [").append(fieldAndFormat.field).append(']');
+            if (fieldAndFormat.matchingPattern != null) {
+                error.append(" which matched [").append(fieldAndFormat.matchingPattern).append(']');
+            }
+            error.append(": ").append(e.getMessage());
+            throw new IllegalArgumentException(error.toString(), e);
         }
-        return new FieldFetcher(fieldContexts, unmappedFieldsFetchAutomaton, unmappedConcreteFields);
+    }
+
+    private static Map<String, FieldContext> buildFieldContexts(
```

Comment:
even though this is private, it would be good adding a few lines of javadoc here. Took me a bit understanding the high level usage of this method